### PR TITLE
Perform lightstep flushing on timer.

### PR DESCRIPTION
**Pull request by RomanDzhabarov**

### CODE REVIEW

File: source/common/tracing/http_tracer_impl.cc [hunk 1]
```diff
@@ -115,27 +115,22 @@ void HttpTracerImpl::populateStats(const Decision& decision) {
 }
 
 LightStepRecorder::LightStepRecorder(const lightstep::TracerImpl& tracer, LightStepSink& sink)
-    : builder_(tracer), sink_(sink) {}
+    : builder_(tracer), sink_(sink) {
+  flush_timer_ = sink_.dispatcher().createTimer([this]() -> void {
+    sink_.tracerStats().timer_flushed_.inc();
+    flushSpans();
+    flush_timer_->enableTimer(std::chrono::milliseconds(1000));
```
Comment (mattklein123) [inline source/common/tracing/http_tracer_impl.cc]: runtime configurable?

File: source/common/tracing/http_tracer_impl.cc [hunk 2]
```diff
@@ -115,27 +115,22 @@ void HttpTracerImpl::populateStats(const Decision& decision) {
 }
 
 LightStepRecorder::LightStepRecorder(const lightstep::TracerImpl& tracer, LightStepSink& sink)
-    : builder_(tracer), sink_(sink) {}
+    : builder_(tracer), sink_(sink) {
+  flush_timer_ = sink_.dispatcher().createTimer([this]() -> void {
+    sink_.tracerStats().timer_flushed_.inc();
+    flushSpans();
```
Comment (mattklein123) [inline source/common/tracing/http_tracer_impl.cc]: check pending spans != 0?

**Comment by mattklein123** [comment]
per offline convo need to use thread local dispatcher for per thread flush timer

**Comment by jmacd** [comment]
I suggest the 1 seconds (timer) and 5 seconds (timeout) constants be made into configuration parameters. Looks good.

File: source/common/tracing/http_tracer_impl.cc (hunk 2 above)
Comment (RomanDzhabarov) [inline source/common/tracing/http_tracer_impl.cc]: good point

File: source/common/tracing/http_tracer_impl.cc [hunk 3]
```diff
@@ -114,16 +114,47 @@ void HttpTracerImpl::populateStats(const Decision& decision) {
...
 
-LightStepRecorder::LightStepRecorder(const lightstep::TracerImpl& tracer, LightStepSink& sink)
-    : builder_(tracer), sink_(sink) {}
+LightStepRecorder::LightStepRecorder(const lightstep::TracerImpl& tracer, LightStepSink& sink,
+                                     Event::Dispatcher& dispatcher)
+    : builder_(tracer), sink_(sink) {
+  flush_timer_ = dispatcher.createTimer([this]() -> void {
+    sink_.tracerStats().timer_flushed_.inc();
+    flushSpans();
+    uint64_t flush_interval =
```
Comment (mattklein123) [inline source/common/tracing/http_tracer_impl.cc]: nit: duplicate code here and below can be collapsed into helper function

Comment (RomanDzhabarov) [inline source/common/tracing/http_tracer_impl.cc]: fixing

File: source/common/tracing/http_tracer_impl.h [hunk 4]
```diff
@@ -110,14 +113,14 @@ class LightStepSink : public HttpSink {
   Runtime::Loader& runtime() { return runtime_; }
   Stats::Store& statsStore() { return stats_store_; }
   LightstepTracerStats& tracerStats() { return tracer_stats_; }
+  Event::Dispatcher& dispatcher() { return dispatcher_; }
```
Comment (mattklein123) [inline source/common/tracing/http_tracer_impl.h]: delete

File: source/server/configuration_impl.cc [hunk 5]
```diff
@@ -88,7 +88,7 @@ void MainImpl::initializeTracers(const Json::Object& tracing_configuration_) {
           opts->guid_generator = [&rand]() { return rand.random(); };
 
           http_tracer_->addSink(Tracing::HttpSinkPtr{new Tracing::LightStepSink(
-              sink.getObject("config"), *cluster_manager_, server_.stats(),
+              sink.getObject("config"), *cluster_manager_, server_.dispatcher(), server_.stats(),
```
Comment (mattklein123) [inline source/server/configuration_impl.cc]: delete

File: source/common/tracing/http_tracer_impl.cc [hunk 6]
```diff
@@ -133,29 +164,24 @@ void LightStepRecorder::RecordSpan(lightstep::collector::Span&& span) {
...
 LightStepSink::LightStepSink(const Json::Object& config, Upstream::ClusterManager& cluster_manager,
-                             Stats::Store& stats, const std::string& service_node,
-                             ThreadLocal::Instance& tls, Runtime::Loader& runtime,
+                             Event::Dispatcher& dispatcher, Stats::Store& stats,
+                             const std::string& service_node, ThreadLocal::Instance& tls,
+                             Runtime::Loader& runtime,
                              std::unique_ptr<lightstep::TracerOptions> options)
     : collector_cluster_(config.getString("collector_cluster")), cm_(cluster_manager),
-      stats_store_(stats),
+      dispatcher_(dispatcher), stats_store_(stats),
```
Comment (RomanDzhabarov) [inline source/common/tracing/http_tracer_impl.cc]: fixed

File: source/common/tracing/http_tracer_impl.h (hunk 4 above)
Comment (RomanDzhabarov) [inline source/common/tracing/http_tracer_impl.h]: fixed

File: source/server/configuration_impl.cc (hunk 5 above)
Comment (RomanDzhabarov) [inline source/server/configuration_impl.cc]: fixed

**Pull request merged**